Document API ability to URL-decode resource names (#12366)
hashicorp/consul: website/content/api-docs/index.mdx

    http://127.0.0.1:8500/v1/kv/foo
```

> ## Parses URL-Encoded Resource Names

Author: I chose the word "Parses" rather than "Accepts" because Consul may still reject the request if the resource name is not valid for that endpoint.

    http://127.0.0.1:8500/v1/kv/foo
```

> ## Accepts URL-Encoded Resource Names

Author: Does this section seem to be in the right place on this page? Or should it be higher/lower?

> ## Accepts URL-Encoded Resource Names

Reviewer: ```suggestion
## URL-Encoded Resource Names
```
Maybe it's just me, but the verb at the beginning of this heading was jarring, especially considering that all (or most) of the other headings in this section have a noun-phrase. Ignore if I'm the odd man out on this one.

Author: I agree. I felt weird about this when I wrote it, too.

> Some Consul HTTP API endpoints require resource names in the URL. To pass a resource
> name containing characters that are invalid in a URL path segment, such as `/` or ` `,
> the resource name must be URL-encoded into the HTTP API call URL.
> 
> Though any resource name *can* be used with the HTTP API if URL-encoded,
> we typically recommend using resource names that don't require URL-encoding anyways.

Reviewer 2: I think the first paragraph isn't really necessary - it was a bug that Consul did not decode correctly but _generally_ people would expect most servers to respect URL-encoded paths

```suggestion
Consul HTTP API endpoints can parse URL-encoded resource names but we typically 
recommend using names that will not require URL-encoding.
```

Author: Though being concise is important, I don't want to be concise at the expense of assuming what users will know or expect. If our assumptions are wrong for just 25% of users, that's still a lot of users who may be confused.

Depending on the validation that Consul applies to a resource name, Consul may still
reject a request if it considers the resource name invalid for that endpoint. And
even if Consul considers the resource name valid, it may degrade other functionality,
such as failed [DNS lookups](/docs/discovery/dns)
for nodes or services with names containing invalid DNS characters.

This HTTP API capability also allows the
[CLI to accept arguments with URL-invalid characters](/commands#accepts-url-invalid-characters).

### Exceptions: Characters that are Always Invalid

The inclusion of certain characters in a URL resource name will cause a request
to be rejected *even if the character is URL-encoded*, including:

> - Linefeed (`%0a`)

Author: @kisunji : could use your eyes for technical accuracy (since you're familiar with this code).

Author: Per Karl, I only gave this its own section for improved scannability

Reviewer: I'm with @kisunji on this one -- especially because there appears to really only be one exception. Maybe change the section heading to something like "Invalid Characters" or something else so that the information would be scannable but topical enough to merit its own section?

Author: Changed to:

---

### Invalid Characters

The linefeed character (`%0a`) will cause a request to be rejected even if it is URL-encoded.
